File: Omnisystem/bootstrap/src/lexer.ts
// Titan lexer — turns source text into a token stream with precise spans.
//
// Handles: line (//) and block (/* */, nested) comments, decimal/hex/bin/oct
// integer and float literals with `_` separators and type suffixes, string
// literals with escapes, char literals, lifetimes ('a), identifiers/keywords,
// and the full operator set (greedy longest-match).

import type { Pos, Span } from './diagnostics.ts';
import { OmniError } from './diagnostics.ts';
import type { Token, TokKind } from './token.ts';
import { KEYWORDS, OPERATORS } from './token.ts';

export class Lexer {
  private src: string;
  private file: string;
  private i = 0;
  private line = 1;
  private col = 1;

  constructor(src: string, file: string) {
    this.src = src;
    this.file = file;
  }

  private pos(): Pos {
    return { offset: this.i, line: this.line, col: this.col };
  }

  private error(msg: string, start: Pos, help: string | null = null): never {
    const sp: Span = { start, end: this.pos() };
    throw new OmniError('lex', msg, sp, this.file, this.src, help);
  }

  private peek(k = 0): string {
    return this.src[this.i + k] ?? '';
  }

  private advance(): string {
    const ch = this.src[this.i++];
    if (ch === '\n') {
      this.line++;
      this.col = 1;
    } else {
      this.col++;
    }
    return ch ?? '';
  }

  private starts(s: string): boolean {
    return this.src.startsWith(s, this.i);
  }

  tokenize(): Token[] {
    const toks: Token[] = [];
    for (;;) {
      this.skipTrivia();
      const start = this.pos();
      if (this.i >= this.src.length) {
        toks.push({ kind: 'eof', value: '', raw: '', span: { start, end: start } });
        return toks;
      }
      const ch = this.peek();
      if (ch === 'r' && this.isRawStringStart()) {
        toks.push(this.lexRawString(start));
      } else if (isIdentStart(ch)) {
        toks.push(this.lexIdent(start));
      } else if (isDigit(ch)) {
        toks.push(this.lexNumber(start));
      } else if (ch === '"') {
        toks.push(this.lexString(start));
      } else if (ch === "'") {
        toks.push(this.lexCharOrLifetime(start));
      } else {
        toks.push(this.lexOperator(start));
      }
    }
  }

  private skipTrivia(): void {
    for (;;) {
      const ch = this.peek();
      if (ch === ' ' || ch === '\t' || ch === '\r' || ch === '\n') {
        this.advance();
      } else if (ch === '/' && this.peek(1) === '/') {
        while (this.i < this.src.length && this.peek() !== '\n') this.advance();
      } else if (ch === '/' && this.peek(1) === '*') {
        this.skipBlockComment();
      } else {
        return;
      }
    }
  }

  private skipBlockComment(): void {
    const start = this.pos();
    this.advance();
    this.advance(); // consume /*
    let depth = 1;
    while (depth > 0) {
      if (this.i >= this.src.length) this.error('unterminated block comment', start);
      if (this.starts('/*')) {
        this.advance();
        this.advance();
        depth++;
      } else if (this.starts('*/')) {
        this.advance();
        this.advance();
        depth--;
      } else {
        this.advance();
      }
    }
  }

  private mk(kind: TokKind, value: string, raw: string, start: Pos): Token {
    return { kind, value, raw, span: { start, end: this.pos() } };
  }

  private lexIdent(start: Pos): Token {
    let s = '';
    while (isIdentContinue(this.peek())) s += this.advance();
    if (s === 'true' || s === 'false') return this.mk('bool', s, s, start);
    if (KEYWORDS.has(s)) return this.mk('keyword', s, s, start);
    return this.mk('ident', s, s, start);
  }

  private lexNumber(start: Pos): Token {
    let raw = '';
    let isFloat = false;
    if (this.peek() === '0' && (this.peek(1) === 'x' || this.peek(1) === 'b' || this.peek(1) === 'o')) {
      raw += this.advance();
      const base = this.advance(); // x/b/o
      raw += base;
      while (isHex(this.peek()) || this.peek() === '_') raw += this.advance();
      const digits = raw.slice(2).replace(/_/g, '');
      const radix = base === 'x' ? 16 : base === 'o' ? 8 : 2;
      const val = parseInt(digits, radix);
      let hsuffix = '';
      while (isIdentContinue(this.peek())) hsuffix += this.advance();
      return this.mk('int', String(val), raw + hsuffix, start);
    }
    while (isDigit(this.peek()) || this.peek() === '_') raw += this.advance();
    if (this.peek() === '.' && isDigit(this.peek(1))) {
      isFloat = true;
      raw += this.advance();
      while (isDigit(this.peek()) || this.peek() === '_') raw += this.advance();
    }
    if (this.peek() === 'e' || this.peek() === 'E') {
      isFloat = true;
      raw += this.advance();
      if (this.peek() === '+' || this.peek() === '-') raw += this.advance();
      while (isDigit(this.peek())) raw += this.advance();
    }
    // optional numeric type suffix (i32, u64, f64, usize, ...)
    let suffix = '';
    while (isIdentContinue(this.peek())) suffix += this.advance();
    if (suffix.startsWith('f')) isFloat = true;
    const clean = raw.replace(/_/g, '');
    return this.mk(isFloat ? 'float' : 'int', clean, raw + suffix, start);
  }

  private lexString(start: Pos): Token {
    this.advance(); // opening quote
    let out = '';
    let raw = '"';
    for (;;) {
      if (this.i >= this.src.length) this.error('unterminated string literal', start);
      const ch = this.advance();
      raw += ch;
      if (ch === '"') break;
      if (ch === '\\') {
        const e = this.advance();
        raw += e;
        if (e === '\n') {
          // Line continuation (Rust string literal semantics): the newline
          // and any leading whitespace on the next line contribute nothing
          // to the string's value, letting multi-line format!()/error
          // strings wrap without embedding a literal newline+indent.
          while (this.peek() === ' ' || this.peek() === '\t') raw += this.advance();
        } else {
          out += this.escape(e, start);
        }
      } else {
        out += ch;
      }
    }
    return this.mk('string', out, raw, start);
  }

  // Rust-style raw string literals: r"..." or r#"..."# / r##"..."## etc. — no
  // escape processing at all, so JSON/regex/paths embedded in Titan specs
  // (e.g. format!(r#"{{"id":"{}"}}"#, ...) in omni-integration/*.titan) don't
  // need every quote and backslash hand-escaped.
  private isRawStringStart(): boolean {
    let k = 1;
    while (this.peek(k) === '#') k++;
    return this.peek(k) === '"';
  }

  private lexRawString(start: Pos): Token {
    let raw = this.advance(); // 'r'
    let hashes = 0;
    while (this.peek() === '#') {
      raw += this.advance();
      hashes++;
    }
    raw += this.advance(); // opening quote
    const closer = '"' + '#'.repeat(hashes);
    let out = '';
    for (;;) {
      if (this.i >= this.src.length) this.error('unterminated raw string literal', start);
      if (this.starts(closer)) {
        for (let k = 0; k < closer.length; k++) raw += this.advance();
        break;
      }
      out += this.advance();
    }
    return this.mk('string', out, raw, start);
  }

  private lexCharOrLifetime(start: Pos): Token {
    // Could be a char literal 'a' / '\n' or a lifetime 'a.
    this.advance(); // opening '
    // lifetime: '<ident> not followed by closing quote
    if (isIdentStart(this.peek())) {
      const save = this.i;
      let name = '';
      while (isIdentContinue(this.peek())) name += this.advance();
      if (this.peek() !== "'") {
        return this.mk('lifetime', name, "'" + name, start);
      }
      // it was a char like 'a' — rewind and fall through
      this.i = save;
      this.col = start.col + 1;
      this.line = start.line;
    }
    let out = '';
    const ch = this.advance();
    if (ch === '\\') {
      out += this.escape(this.advance(), start);
    } else {
      out += ch;
    }
    if (this.peek() !== "'") this.error("unterminated char literal", start, "char literals hold exactly one character, e.g. 'a'");
    this.advance();
    return this.mk('char', out, "'" + out + "'", start);
  }

  private escape(e: string, start: Pos): string {
    switch (e) {
      case 'n': return '\n';
      case 't': return '\t';
      case 'r': return '\r';
      case '0': return '\0';
      case '\\': return '\\';
      case '"': return '"';
      case "'": return "'";
      default: this.error(`unknown escape sequence \\${e}`, start);
    }
  }

  private lexOperator(start: Pos): Token {
    for (const op of OPERATORS) {
      if (this.starts(op)) {
        for (let k = 0; k < op.length; k++) this.advance();
        return this.mk('op', op, op, start);
      }
    }
    const bad = this.advance();
    this.error(`unexpected character '${bad}'`, start);
  }
}

function isDigit(c: string): boolean {
  return c >= '0' && c <= '9';
}
function isHex(c: string): boolean {
  return isDigit(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}
function isIdentStart(c: string): boolean {
  return c === '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}
function isIdentContinue(c: string): boolean {
  return isIdentStart(c) || isDigit(c);
}

File: Omnisystem/bootstrap-rs/src/lexer.rs
//! Titan lexer — source text to a token stream with precise spans.
//!
//! Handles line and nested block comments, decimal/hex/bin/oct integers and
//! floats with `_` separators and type suffixes, string/char literals with
//! escapes, lifetimes, identifiers/keywords, and greedy longest-match
//! operators.

use crate::diag::{OResult, OmniError, Phase, Pos, Span};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokKind {
    Int,
    Float,
    Str,
    Char,
    Bool,
    Ident,
    Keyword,
    Lifetime,
    Op,
    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokKind,
    pub value: String,
    pub span: Span,
}

pub const KEYWORDS: &[&str] = &[
    "use", "pub", "mod", "struct", "enum", "impl", "trait", "fn", "let", "mut", "const", "static",
    "if", "else", "match", "while", "for", "in", "loop", "break", "continue", "return", "self",
    "Self", "true", "false", "as", "where", "ref", "move", "dyn", "async", "await", "type",
    "unsafe", "extern",
];

/// Multi-char operators first so matching is greedy.
pub const OPERATORS: &[&str] = &[
    "..=", "...", "<<=", ">>=", "->", "=>", "::", "==", "!=", "<=", ">=", "&&", "||", "+=", "-=",
    "*=", "/=", "%=", "&=", "|=", "^=", "<<", ">>", "..", "+", "-", "*", "/", "%", "=", "<", ">",
    "!", "&", "|", "^", "~", "?", ".", ",", ";", ":", "(", ")", "{", "}", "[", "]", "@", "#",
];

pub struct Lexer<'a> {
    src: Vec<char>,
    file: &'a str,
    i: usize,
    line: u32,
    col: u32,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str, file: &'a str) -> Self {
        Lexer { src: src.chars().collect(), file, i: 0, line: 1, col: 1 }
    }

    fn pos(&self) -> Pos {
        Pos { line: self.line, col: self.col }
    }

    fn err(&self, msg: impl Into<String>, start: Pos) -> OmniError {
        OmniError::new(Phase::Lex, msg, Span { start, end: self.pos() }, self.file)
    }

    fn peek(&self, k: usize) -> char {
        self.src.get(self.i + k).copied().unwrap_or('\0')
    }

    fn advance(&mut self) -> char {
        let c = self.peek(0);
        self.i += 1;
        if c == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        c
    }

    fn starts(&self, s: &str) -> bool {
        let rest = &self.src[self.i.min(self.src.len())..];
        s.chars().enumerate().all(|(k, c)| rest.get(k) == Some(&c))
    }

    pub fn tokenize(mut self) -> OResult<Vec<Token>> {
        let mut toks = Vec::new();
        loop {
            self.skip_trivia()?;
            let start = self.pos();
            if self.i >= self.src.len() {
                toks.push(Token { kind: TokKind::Eof, value: String::new(), span: Span { start, end: start } });
                return Ok(toks);
            }
            let c = self.peek(0);
            let tok = if c == 'r' && (self.peek(1) == '"' || (self.peek(1) == '#' && (self.peek(2) == '"' || self.peek(2) == '#'))) {
                self.lex_raw_string(start)?
            } else if c == 'b' && self.peek(1) == '"' {
                self.advance(); // consume 'b' — byte strings lex as ordinary strings
                self.lex_string(start)?
            } else if c == 'b' && self.peek(1) == '\'' {
                // byte char b'E' — lexes as an integer (the byte value)
                self.advance();
                let t = self.lex_char_or_lifetime(start)?;
                let byte = t.value.chars().next().map(|ch| ch as u32).unwrap_or(0);
                Token { kind: TokKind::Int, value: byte.to_string(), span: t.span }
            } else if c == '_' || c.is_ascii_alphabetic() {
                self.lex_ident(start)
            } else if c.is_ascii_digit() {
                self.lex_number(start)?
            } else if c == '"' {
                self.lex_string(start)?
            } else if c == '\'' {
                self.lex_char_or_lifetime(start)?
            } else {
                self.lex_operator(start)?
            };
            toks.push(tok);
        }
    }

    fn skip_trivia(&mut self) -> OResult<()> {
        loop {
            let c = self.peek(0);
            if c == ' ' || c == '\t' || c == '\r' || c == '\n' {
                self.advance();
            } else if c == '/' && self.peek(1) == '/' {
                while self.i < self.src.len() && self.peek(0) != '\n' {
                    self.advance();
                }
            } else if c == '/' && self.peek(1) == '*' {
                let start = self.pos();
                self.advance();
                self.advance();
                let mut depth = 1u32;
                while depth > 0 {
                    if self.i >= self.src.len() {
                        return Err(self.err("unterminated block comment", start));
                    }
                    if self.starts("/*") {
                        self.advance();
                        self.advance();
                        depth += 1;
                    } else if self.starts("*/") {
                        self.advance();
                        self.advance();
                        depth -= 1;
                    } else {
                        self.advance();
                    }
                }
            } else {
                return Ok(());
            }
        }
    }

    fn mk(&self, kind: TokKind, value: String, start: Pos) -> Token {
        Token { kind, value, span: Span { start, end: self.pos() } }
    }

    fn lex_ident(&mut self, start: Pos) -> Token {
        let mut s = String::new();
        while is_ident_continue(self.peek(0)) {
            s.push(self.advance());
        }
        if s == "true" || s == "false" {
            return self.mk(TokKind::Bool, s, start);
        }
        if KEYWORDS.contains(&s.as_str()) {
            return self.mk(TokKind::Keyword, s, start);
        }
        self.mk(TokKind::Ident, s, start)
    }

    fn lex_number(&mut self, start: Pos) -> OResult<Token> {
        // hex / bin / oct
        if self.peek(0) == '0' && matches!(self.peek(1), 'x' | 'b' | 'o') {
            self.advance();
            let base = self.advance();
            let mut digits = String::new();
            while self.peek(0).is_ascii_alphanumeric() || self.peek(0) == '_' {
                let c = self.advance();
                if c != '_' {
                    digits.push(c);
                }
            }
            // strip a numeric type suffix like u8/i32 from the digit run
            let radix = match base { 'x' => 16, 'o' => 8, _ => 2 };
            let cut = digits
                .char_indices()
                .find(|(_, ch)| !ch.is_digit(radix))
                .map(|(idx, _)| idx)
                .unwrap_or(digits.len());
            let (num, _suffix) = digits.split_at(cut);
            let val = i64::from_str_radix(num, radix)
                .map_err(|_| self.err(format!("invalid numeric literal `{digits}`"), start))?;
            return Ok(self.mk(TokKind::Int, val.to_string(), start));
        }
        let mut raw = String::new();
        let mut is_float = false;
        while self.peek(0).is_ascii_digit() || self.peek(0) == '_' {
            let c = self.advance();
            if c != '_' {
                raw.push(c);
            }
        }
        if self.peek(0) == '.' && self.peek(1).is_ascii_digit() {
            is_float = true;
            raw.push(self.advance());
            while self.peek(0).is_ascii_digit() || self.peek(0) == '_' {
                let c = self.advance();
                if c != '_' {
                    raw.push(c);
                }
            }
        }
        if matches!(self.peek(0), 'e' | 'E') && (self.peek(1).is_ascii_digit() || (matches!(self.peek(1), '+' | '-') && self.peek(2).is_ascii_digit())) {
            is_float = true;
            raw.push(self.advance());
            if matches!(self.peek(0), '+' | '-') {
                raw.push(self.advance());
            }
            while self.peek(0).is_ascii_digit() {
                raw.push(self.advance());
            }
        }
        // type suffix (i32, u8, f64, usize, ...)
        let mut suffix = String::new();
        while is_ident_continue(self.peek(0)) {
            suffix.push(self.advance());
        }
        if suffix.starts_with('f') {
            is_float = true;
        }
        Ok(self.mk(if is_float { TokKind::Float } else { TokKind::Int }, raw, start))
    }

    /// r"..." / r#"..."# / r##"..."## — no escapes, closes on `"` + same # count.
    fn lex_raw_string(&mut self, start: Pos) -> OResult<Token> {
        self.advance(); // r
        let mut hashes = 0usize;
        while self.peek(0) == '#' {
            self.advance();
            hashes += 1;
        }
        if self.peek(0) != '"' {
            return Err(self.err("expected '\"' to open raw string", start));
        }
        self.advance();
        let closer: String = format!("\"{}", "#".repeat(hashes));
        let mut out = String::new();
        loop {
            if self.i >= self.src.len() {
                return Err(self.err("unterminated raw string literal", start));
            }
            if self.starts(&closer) {
                for _ in 0..closer.len() {
                    self.advance();
                }
                break;
            }
            out.push(self.advance());
        }
        Ok(self.mk(TokKind::Str, out, start))
    }

    fn lex_string(&mut self, start: Pos) -> OResult<Token> {
        self.advance(); // opening "
        let mut out = String::new();
        loop {
            if self.i >= self.src.len() {
                return Err(self.err("unterminated string literal", start));
            }
            let c = self.advance();
            if c == '"' {
                break;
            }
            if c == '\\' {
                let e = self.advance();
                out.push(self.escape(e, start)?);
            } else {
                out.push(c);
            }
        }
        Ok(self.mk(TokKind::Str, out, start))
    }

    fn lex_char_or_lifetime(&mut self, start: Pos) -> OResult<Token> {
        self.advance(); // opening '
        // lifetime: '<ident> not closed by another quote
        if is_ident_start(self.peek(0)) {
            let save = (self.i, self.line, self.col);
            let mut name = String::new();
            while is_ident_continue(self.peek(0)) {
                name.push(self.advance());
            }
            if self.peek(0) != '\'' {
                return Ok(self.mk(TokKind::Lifetime, name, start));
            }
            // it was a char literal like 'a' — rewind
            (self.i, self.line, self.col) = save;
        }
        let c = self.advance();
        let ch = if c == '\\' {
            let e = self.advance();
            self.escape(e, start)?
        } else {
            c
        };
        if self.peek(0) != '\'' {
            return Err(self
                .err("unterminated char literal", start)
                .with_help("char literals hold exactly one character, e.g. 'a'"));
        }
        self.advance();
        Ok(self.mk(TokKind::Char, ch.to_string(), start))
    }

    fn escape(&self, e: char, start: Pos) -> OResult<char> {
        Ok(match e {
            'n' => '\n',
            't' => '\t',
            'r' => '\r',
            '0' => '\0',
            '\\' => '\\',
            '"' => '"',
            '\'' => '\'',
            other => return Err(self.err(format!("unknown escape sequence \\{}", other), start)),
        })
    }

    fn lex_operator(&mut self, start: Pos) -> OResult<Token> {
        for op in OPERATORS {
            if self.starts(op) {
                for _ in 0..op.len() {
                    self.advance();
                }
                return Ok(self.mk(TokKind::Op, (*op).to_string(), start));
            }
        }
        let bad = self.advance();
        Err(self.err(format!("unexpected character '{bad}'"), start))
    }
}

fn is_ident_start(c: char) -> bool {
    c == '_' || c.is_ascii_alphabetic()
}
fn is_ident_continue(c: char) -> bool {
    is_ident_start(c) || c.is_ascii_digit()
}
